File: server/src/main/java/server/models/DocumentHistory.java
package server.models;

import jakarta.persistence.*;
import lombok.*;
import server.models.enums.DocumentHistoryAction;
import server.models.enums.DocumentStatus;
import server.models.enums.DocumentType;

import java.time.LocalDateTime;

@Entity
@Table(name = "document_histories")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class DocumentHistory {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    // Tài liệu gốc
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "document_id", nullable = false)
    private Document document;

    // Ai tạo snapshot (MANAGER khi NOTE, SECRETARY khi UPDATE)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by_id", nullable = false)
    private Account createdBy;

    // Hành động
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DocumentHistoryAction action;

    // Version tăng dần theo từng lần ghi lịch sử
    @Column(nullable = false)
    private Integer version;

    // Snapshot nội dung
    private String title;

    @Column(columnDefinition = "TEXT")
    private String content;

    private String fileUrl;

    // PROJECT fields
    private String projectName;

    @Column(columnDefinition = "TEXT")
    private String projectDescription;

    private String projectPriority;   // lưu chuỗi cho đơn giản

    private String projectDeadline;   // yyyy-MM-dd

    // ADMINISTRATIVE fields
    private String fundName;

    private Double fundBalance;

    @Column(columnDefinition = "TEXT")
    private String fundPurpose;

    @Enumerated(EnumType.STRING)
    private DocumentType type;

    @Enumerated(EnumType.STRING)
    private DocumentStatus status;

    @Column(columnDefinition = "TEXT")
    private String signature;

    // Note tại thời điểm snapshot (nếu là NOTE)
    @Column(columnDefinition = "TEXT")
    private String managerNote;

    private LocalDateTime createdAt;
}

File: mobile/lib/models/document_history.dart
import 'package:mobile/models/account.dart';

class DocumentHistory {
  final int id;
  final int version;
  final String action;
  final Account? createdBy;
  final String? title;
  final String? content;
  final String? fileUrl;

  // Project snapshot
  final String? projectName;
  final String? projectDescription;
  final String? projectPriority;
  final String? projectDeadline;

  // Fund snapshot
  final String? fundName;
  final double? fundBalance;
  final String? fundPurpose;

  final String? type;
  final String? status;
  final String? signature;
  final String? managerNote;
  final DateTime? createdAt;

  DocumentHistory({
    required this.id,
    required this.version,
    required this.action,
    this.createdBy,
    this.title,
    this.content,
    this.fileUrl,
    this.projectName,
    this.projectDescription,
    this.projectPriority,
    this.projectDeadline,
    this.fundName,
    this.fundBalance,
    this.fundPurpose,
    this.type,
    this.status,
    this.signature,
    this.managerNote,
    this.createdAt,
  });

  factory DocumentHistory.fromJson(Map<String, dynamic> json) {
    int _toInt(dynamic v) {
      if (v == null) return 0;
      if (v is int) return v;
      return int.tryParse(v.toString()) ?? 0;
    }

    double? _toDouble(dynamic v) {
      if (v == null) return null;
      if (v is num) return v.toDouble();
      return double.tryParse(v.toString());
    }

    DateTime? _toDate(dynamic v) {
      if (v == null) return null;
      try { return DateTime.parse(v.toString()); } catch (_) { return null; }
    }

    return DocumentHistory(
      id: _toInt(json['id']),
      version: _toInt(json['version']),
      action: json['action']?.toString() ?? '',
      createdBy: json['createdBy'] != null ? Account.fromJson(json['createdBy']) : null,
      title: json['title']?.toString(),
      content: json['content']?.toString(),
      fileUrl: json['fileUrl']?.toString(),
      projectName: json['projectName']?.toString(),
      projectDescription: json['projectDescription']?.toString(),
      projectPriority: json['projectPriority']?.toString(),
      projectDeadline: json['projectDeadline']?.toString(),
      fundName: json['fundName']?.toString(),
      fundBalance: _toDouble(json['fundBalance']),
      fundPurpose: json['fundPurpose']?.toString(),
      type: json['type']?.toString(),
      status: json['status']?.toString(),
      signature: json['signature']?.toString(),
      managerNote: json['managerNote']?.toString(),
      createdAt: _toDate(json['createdAt']),
    );
  }
}
